plumbing/format/packfile/decoder.go

```diff
@@ -2,9 +2,11 @@ package packfile
 
```

Comment:
shall we delete this file?

Reply:
Packfile must work always with an index, if in some place a Packfile without an index is needed, we need to evaluate why.

The index is created on the cloning process and when a new packfile is downloaded, so it's always there.

Could you point me to places that Decoder is used without the possibility to get the index?

Reply:
https://github.com/src-d/go-git/blob/master/storage/filesystem/object.go#L463-L490

As you can see, `newPackFileIter` receives an `*idxfile.Index` argument, but it's always nil from `NewPackfileIter`

I mean, if we're 100 % sure there will always be an index we can safely delete this and move all the tests from Decoder to Packfile.

Another feature we don't have in the Packfile, whereas we do in Decoder is filling ObjectStorages from the Packfile. I'm not sure that's used anywhere, though.

Reply:
Update: yes, it's used for this exact thing, in this case, again without index: https://github.com/src-d/go-git/blob/0c2618bc7d495322c6413b0ff62549712747cf9e/plumbing/format/packfile/common.go#L26-L41

Here it's used to generate the index: https://github.com/src-d/go-git/blob/ae788cfabbc02c2f836f5d8c3cc18021a97e9a88/storage/filesystem/dotgit/writers.go#L58-L72

Reply:
the problem with that usages is that we are putting all the objects on memory... and is not the solution.

Per example, in the case of `UpdateObjectStorage` we just need to copy packfile on the .git folder using `writePackfileToObjectStorage`. Storer must implement PackfileWriter always.

The second case is where the index is built. That's is the perfect case to use the packfile parser object.